### ISSUE TITLE
Final achors of trained model

### ISSUE THREAD

**Issue by RobinBram** [ultralytics/yolov5#2338]
Is there a way to see which anchors are used in the final model I have trained with autoanchor?

**Comment by glenn-jocher** [comment]
@RobinBram YOLOv5 anchors are saved as [Detect()](https://github.com/ultralytics/yolov5/blob/fab5085674f7748dc16d7ca25afb225fa441bc9d/models/yolo.py#L22) layer attributes on model creation, and updated as necessary by  autoanchor evolution before training starts. Their exact location is here:
https://github.com/ultralytics/yolov5/blob/fab5085674f7748dc16d7ca25afb225fa441bc9d/models/yolo.py#L34-L35

You can examine the anchors of any trained YOLOv5 model like this:

### Input
```python
import torch

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s', pretrained=True, autoshape=False)  # hub model
# -- or --
model = torch.load('yolov5s.pt')['model']  # local model

# Anchors
m = model.model[-1]  # Detect() layer
print(m.anchor_grid.squeeze())  # print anchors
```

### Output
```python
#           x     y
tensor([[[ 10.,  13.],
         [ 16.,  30.],
         [ 33.,  23.]],  # P3/8-small

        [[ 30.,  61.],
         [ 62.,  45.],
         [ 59., 119.]],  # P4/16-medium

        [[116.,  90.],
         [156., 198.],
         [373., 326.]]], dtype=torch.float16)  # P5/32-large
```

**Comment by RobinBram** [comment]
@glenn-jocher Great thank you! However, I'm confused about the acnhors that i get. I traindd the model with anchor_t: 3.32
anchors: 2.0 which I got from evolve, but I know get the following 3! anchors from the outout from the code. 

tensor([[[ 10.,  13.],
         [ 16.,  30.],
         [ 33.,  23.]],

        [[ 30.,  61.],
         [ 62.,  45.],
         [ 59., 119.]],

        [[116.,  90.],
         [156., 198.],
         [373., 326.]]], device='cuda:0')

**Comment by RobinBram** [comment]
Great! Thanks! I am however confused by the results i get. I trained with anchor_t: 3.32 and anchors: 2.0 which I got from evolve but I get the following 3! anchors from the code(which seems to be the default anchors): 

tensor([[[ 10.,  13.],
         [ 16.,  30.],
         [ 33.,  23.]],

        [[ 30.,  61.],
         [ 62.,  45.],
         [ 59., 119.]],

        [[116.,  90.],
         [156., 198.],
         [373., 326.]]], device='cuda:0')

This is what i ran:
import torch 
model=torch.load('runs/train/exp2/weights/best.pt')['model']
m = model.model[-1]
print(m.anchor_grid.squeeze())

**Comment by glenn-jocher** [comment]
@RobinBram your model is using the default anchors. You can set model yaml `anchors: 2` to use 2 anchors instead of 3.

You can also uncomment the anchors field in your hyp file as well, which will override anchors defined in your model yaml. Make sure you git pull to get the latest code as well, we just had a recent anchor related PR.

**Comment by RobinBram** [comment]
'''# Hyperparameter Evolution Results
# Generations: 301
# Metrics:      0.952      0.96     0.985       0.8    0.0196   0.00509         0

lr0: 0.011
lrf: 0.22
momentum: 0.852
weight_decay: 0.00052
warmup_epochs: 1.83
warmup_momentum: 0.852
warmup_bias_lr: 0.0958
box: 0.0568
cls: 0.493
cls_pw: 1.01
obj: 0.994
obj_pw: 1.22
iou_t: 0.2
anchor_t: 3.32
anchors: 2.0
fl_gamma: 1.0
hsv_h: 0.015
hsv_s: 0.635
hsv_v: 0.291
degrees: 2.08
translate: 0.111
scale: 0.48
shear: 0.973
perspective: 0.0005
flipud: 0.542
fliplr: 0.5
mosaic: 0.106
mixup: 0.115
'''
This is what my hyp-file looked like for training which I got from running evolve for 300 epochs. I suspect the anchors parameter is not taken into account, (because it whould wright over the existing anchors right?). 

When running evolve, the anchor parameter took non integer values, which is kind of weird. Not shure that anchor_t did anything then either. I'm also wondering why iou_t, fl_gamma, perspective and fliplr did not evolve. @glenn-jocher 
![evolve](https://user-images.githubusercontent.com/70885556/109934082-64b9c180-7ccc-11eb-80d8-965d5180aa7e.png)

**Comment by glenn-jocher** [comment]
@RobinBram anchors are rounded to the nearest integer when used:
https://github.com/ultralytics/yolov5/blob/e931b9da33f45551928059b8d61bddd50e401e48/models/yolo.py#L82

Hyperparameters that start at zero do not vary. Also the `meta` dictionary in train.py manages hyperparameter limits, setting some to zero variation, such as fl_gamma and perspective. 

Your results do not vary greatly in general, even the worst generation performed well, which is strange. Not sure how to interpret.

**Comment by RobinBram** [comment]
Great thank you! The mutation-scale were 0 for the hyperparameters that did not evolve. The evolution was performed on a well trained model for 10 epochs each on a pretty easy dataset. (map @0.5 over 0.98) When retraining the model in full with the new hyperparameters, the scores went up further, but maybe this is just overfitting to the validation set. It will show later. 

I still don't think I've gotten why the anchors: 2.0 in the hyp.yaml file did not work. As I said, the model still has the default anchors. Will try again with git pull. Last pull was about a month ago. 

If i change in the model.yaml file, do I change 


`anchors:
  - [10,13, 16,30, 33,23]  # P3/8
  - [30,61, 62,45, 59,119]  # P4/16
  - [116,90, 156,198, 373,326]  # P5/32`
to 
`anchors: 2`
?

**Comment by RobinBram** [comment]
Anchors seem to work properly now after git pull.

**Comment by glenn-jocher** [comment]
@RobinBram from your evolution it seems 2 anchors work best (though the lower limit is set at 2, so its possible if you relax it to 1, then evolution might go towards 1 anchor). The evolution does not provide you with these anchor values though, it simply runs autoanchor on every generation for the rounded anchor hyperparameter, so to use these hyps you simply point to them when training a new model with `python train.py --hyp hyp.evolved.yaml`, and let autoanchor again solve for 2 anchors.
